### PR TITLE
location fromstring

### PR DESCRIPTION
**Pull request by mdehoon**
This PR adds an abstract base class `Location` to `SimpleLocation` and `CompoundLocation` in `Bio.SeqFeature`, with
a `fromstring` class method to create locations. This PR supersedes PR #4063 .

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by mdehoon** [comment]
See also #4130 , which needs to be addressed first.

File: Bio/GenBank/__init__.py [hunk 1]
```diff
@@ -62,380 +65,6 @@
...
-    "join(complement(149815..150200),complement(293787..295573),NC_016402.1:6618..6676,181647..181905)"
-)
-assert not _re_simple_location.match(
-    "join(complement(149815..150200),complement(293787..295573),NC_016402.1:6618..6676,181647..181905)"
-)
-
-_solo_bond = re.compile(r"bond\(%s\)" % _solo_location)
-assert _solo_bond.match("bond(196)")
-assert _solo_bond.search("bond(196)")
-assert _solo_bond.search("join(bond(284),bond(305),bond(309),bond(305))")
```
Comment (peterjc) [inline Bio/GenBank/__init__.py:180]: Discussion on https://github.com/biopython/biopython/issues/3600 suggests removing the asserts (which started as regex documentation).

We could move them into the test file (as unittests), but that leaves the question of how best to documentation the regexes?

Comment (mdehoon) [reply Bio/GenBank/__init__.py:180]: I think keeping the `assert` statements here help with code readability. How to handle this case is something to be considered as part of #3600  . I have left a comment there.

**Comment by mdehoon** [comment]
@peterjc  Can this be merged?

**Comment by peterjc** [comment]
Did you manage to run some performance benchmarks? The features are a big part of the processing time on large GenBank files, so it would be reassuring if this has not slowed things down.

File: Bio/GenBank/__init__.py [hunk 2]
```diff
@@ -482,18 +111,12 @@ def __iter__(self):
...
 
 
-class ParserFailureError(Exception):
+class ParserFailureError(ValueError):
     """Failure caused by some kind of problem in the parser."""
 
     pass
 
 
-class LocationParserError(Exception):
```
Comment (peterjc) [inline Bio/GenBank/__init__.py:491]: Removing this, or rather moving it under ``Bio/SeqFeature.py`` is logical, but could break some people's scripts.

Can we leave a stub here too? e.g. ``from Bio.SeqFeature import LocationParserError`` with a comment about it being for backward compatibility?

File: Bio/GenBank/__init__.py [hunk 3]
```diff
@@ -45,7 +42,13 @@
...
 from Bio import BiopythonParserWarning
 from Bio.Seq import Seq
-from Bio import SeqFeature
+from Bio.SeqFeature import (
+    SimpleLocation,
+    Location,
+    Reference,
+    SeqFeature,
+    LocationParserError,
+)
```
Comment (peterjc) [inline Bio/GenBank/__init__.py:51]: Minor: Style wise although not formalised we've avoid the bracket multiline imports, I would use multiple import lines here - sorted alphabetically.

(There are some issues about automatically sorting/formatting our imports)

File: Bio/SeqFeature.py [hunk 4]
```diff
@@ -684,7 +776,132 @@ def __eq__(self, other):
...
+class Location(ABC):
+    """Abstract base class representing a location."""
+
+    @abstractmethod
+    def __repr__(self):
+        """Represent the Location object as a string for debugging."""
+        return f"{self.__class__.__name__}(...)"
+
+    def fromstring(location_line, length=None, circular=False, stranded=True):
+        """Create a Location object from a string.
```
Comment (peterjc) [inline Bio/SeqFeature.py:788]: I think the docstring should including something like:

```
This should accept any valid location string in the INSDC Feature Table format
(https://www.insdc.org/submitting-standards/feature-table/) as used in GenBank,
DDBJ and EMBL files.
```

File: Bio/SeqFeature.py [hunk 5]
```diff
@@ -684,7 +776,132 @@ def __eq__(self, other):
...
+
+        You can optionally provide a reference name:
+
+        >>> Location.fromstring("AL391218.9:105173..108462", 2000000)
+        SimpleLocation(ExactPosition(105172), ExactPosition(108462), strand=1, ref='AL391218.9')
+
+        >>> Location.fromstring("<2644..159", 2868, "circular")
+        CompoundLocation([SimpleLocation(BeforePosition(2643), ExactPosition(2868), strand=1), SimpleLocation(ExactPosition(0), ExactPosition(159), strand=1)], 'join')
+        """
+        if location_line.startswith("complement"):
```
Comment (peterjc) [inline Bio/SeqFeature.py:830]: The original checked for the brackets:

```python
        if location_line.startswith("complement("):
            assert location_line.endswith(")")
            location_line = location_line[11:-1]
            strand = -1
        ...
```

Checking the opening bracket is present too should make trivial difference to speed (and would be consistent with the join etc changes below).

The closing bracket check if performed ought not to be an assert, but raise an explicit error?

**Comment by mdehoon** [comment]
> Did you manage to run some performance benchmarks? The features are a big part of the processing time on large GenBank files, so it would be reassuring if this has not slowed things down.

Yes. The original version was a bit funny because it used `match` on the regular expression to confirm that the string was consistent with the pattern, but then threw away the result. The version in this PR uses string processing instead of regular expressions where possible, as the regular expressions were significantly slower.

File: Bio/GenBank/__init__.py (hunk 2 above)
Comment (mdehoon) [reply Bio/GenBank/__init__.py:491]: It's already there, as `LocationParserError` is still needed in `Bio/GenBank/__init__.py`.

File: Bio/GenBank/__init__.py (hunk 3 above)
Comment (mdehoon) [reply Bio/GenBank/__init__.py:51]: done

File: Bio/SeqFeature.py (hunk 4 above)
Comment (mdehoon) [reply Bio/SeqFeature.py:788]: done

File: Bio/SeqFeature.py (hunk 5 above)
Comment (mdehoon) [reply Bio/SeqFeature.py:830]: done

**Comment by mdehoon** [comment]
> Did you manage to run some performance benchmarks? The features are a big part of the processing time on large GenBank files, so it would be reassuring if this has not slowed things down.

Some timings parsing the GenBank protein annotations for hg38:
Current code in master: 111.38 seconds
After this PR: 111.67 seconds
As a test, in this PR I replaced the regular expression to split a line by commas by `str.split` if the line does not contain a `one-of` location; then parsing takes 110.91 seconds. So we can speed up a bit by avoiding the regular expressions, at the cost of making the code more complex, but the gains will be marginal.

**Review by peterjc — APPROVED** [review]
I suggest a squash-and-merge with a re-written commit message.

**Comment by mdehoon** [comment]
Thanks for looking at this

**Pull request merged**